source/common/router/scoped_rds.cc

```diff
@@ -100,72 +107,205 @@ ScopedRdsConfigSubscription::ScopedRdsConfigSubscription(
   });
 }
 
+ScopedRdsConfigSubscription::RdsRouteConfigProviderHelper::RdsRouteConfigProviderHelper(
+    ScopedRdsConfigSubscription& parent, std::string scope_name,
+    envoy::config::filter::network::http_connection_manager::v2::Rds& rds,
+    Init::Manager& init_manager)
+    : parent_(parent), scope_name_(scope_name),
+      route_provider_(static_cast<RdsRouteConfigProviderImpl*>(
+          parent_.route_config_provider_manager_
+              .createRdsRouteConfigProvider(rds, parent_.factory_context_, parent_.stat_prefix_,
+                                            init_manager)
+              .release())),
+      rds_update_callback_handle_(route_provider_->subscription().addUpdateCallback([this]() {
+        // Subscribe to RDS update.
+        parent_.onRdsConfigUpdate(scope_name_, route_provider_->subscription());
+      })) {}
+
 void ScopedRdsConfigSubscription::onConfigUpdate(
-    const Protobuf::RepeatedPtrField<ProtobufWkt::Any>& resources,
+    const Protobuf::RepeatedPtrField<envoy::api::v2::Resource>& added_resources,
+    const Protobuf::RepeatedPtrField<std::string>& removed_resources,
     const std::string& version_info) {
-  std::vector<envoy::api::v2::ScopedRouteConfiguration> scoped_routes;
-  for (const auto& resource_any : resources) {
-    scoped_routes.emplace_back(MessageUtil::anyConvert<envoy::api::v2::ScopedRouteConfiguration>(
-        resource_any, validation_visitor_));
+  bool any_applied = false;
+  std::vector<std::string> exception_msgs;
+  absl::flat_hash_set<std::string> unique_resource_names;
+  envoy::config::filter::network::http_connection_manager::v2::Rds rds;
+  rds.mutable_config_source()->MergeFrom(rds_config_source_);
+
+  // If new route config sources come after the factory_context_.initManager()'s initialize() been
+  // called, that initManager can't accept new targets. Instead we use a local override which will
+  // start new subscriptions but not wait on them to be ready.
+  std::unique_ptr<Init::ManagerImpl> overriding_init_manager;
```

Comment:
Thanks for the lifetime/ownership cleanup, this is much easier to grok now. I'm still wondering if there is anything we can do to avoid this pseudo-init manager from appearing? It works AFAICT, but it's a bit of a conceptual misdirect; i.e. why is there init manager happenings occurring once we're active?

Reply:
Maybe we can loop in @mergeconflict and have an offline discussion on this as well. It's the main place that I'm seeing things that seem a stretch from normal Envoy convention below.

Reply:
Happy to have offline chat. 
to your first question, I *think* the init-manager was in almost every xDS's subscription creation because of Envoy's history: we want to have some sort of coordination between different components' kicking off, usually it's "waiting for a listener to be warmed up".
I think VHDS and SRDS has advanced Envoy to a point that some new resource subscriptions are more "dynamic", that they can be created beyond the barrier of Listener.